### PR TITLE
Wait for enough chain confirmations on client start for archiving and signing

### PR DESCRIPTION
**Pull request by nkuba**
In this PR we implement chain state confirmations checks on client start. We cover two cases when the client starts:
- keep is inactive and we archive its' data
- keep is awaiting signatures and we start signing process.

~~Depends on: https://github.com/keep-network/keep-ecdsa/pull/372~~
Closes: https://github.com/keep-network/keep-ecdsa/issues/371

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -279,6 +348,43 @@ func checkAwaitingSignature(
...
+		isStillAwaitingSignature, err := waitForChainConfirmation(
+			ethereumChain,
+			currentBlock,
+			func() (bool, error) {
+				return ethereumChain.IsAwaitingSignature(keepAddress, latestDigest)
+			},
+		)
+		if err != nil {
+			logger.Errorf(
+				"failed to confirm signing request for digest [%+x] and keep [%s]",
```
Comment (lukasz-zimnoch) [inline pkg/client/client.go:372]: Maybe we can log `err` as well?

File: pkg/client/client.go [hunk 2]
```diff
@@ -279,6 +348,43 @@ func checkAwaitingSignature(
...
 	if isAwaitingDigest {
+		if err := requestedSignatures.add(keepAddress, latestDigest); err != nil {
+			logger.Errorf("failed to register signature request [%v]", err)
+			return
+		}
+		defer requestedSignatures.remove(keepAddress, latestDigest)
+
+		currentBlock, err := ethereumChain.BlockCounter().CurrentBlock()
+		if err != nil {
+			logger.Errorf("failed to get current block height [%v]", err)
```
Comment (lukasz-zimnoch) [inline pkg/client/client.go:359]: Same with the colon here.

File: pkg/client/client.go [hunk 3]
```diff
@@ -279,6 +348,43 @@ func checkAwaitingSignature(
 	}
 
 	if isAwaitingDigest {
+		if err := requestedSignatures.add(keepAddress, latestDigest); err != nil {
+			logger.Errorf("failed to register signature request [%v]", err)
```
Comment (lukasz-zimnoch) [inline pkg/client/client.go:352]: If I'm not wrong we have a convention to add a `:` before we print `err`. So I think it should be: `logger.Errorf("failed to register signature request: [%v]", err)`

File: pkg/client/client.go [hunk 4]
```diff
@@ -223,32 +283,40 @@ func monitorSigningRequests(
...
+					ethereumChain,
+					event.BlockNumber,
+					func() (bool, error) {
+						return ethereumChain.IsAwaitingSignature(keepAddress, event.Digest)
+					},
 				)
-				return
-			}
+				if err != nil {
+					logger.Errorf(
```
Comment (lukasz-zimnoch) [inline pkg/client/client.go:301]: Can we log the cause (`err`)?

File: pkg/client/client.go [hunk 5]
```diff
@@ -45,28 +46,78 @@ func Initialize(
...
+					return
+				}
+				logger.Warningf("keep [%s] is still active", keepAddress.String())
+			}
+
 			signers, err := keepsRegistry.GetSigners(keepAddress)
 			if err != nil {
+				// If there are no signers for loaded keep that something is clearly
+				// wrong. We don't want to continue processing for this keep.
 				logger.Errorf("no signers for keep [%s]", keepAddress.String())
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/client/client.go:119]: I second that comment.

File: pkg/client/client.go (hunk 3 above)
Comment (pdyraga) [reply pkg/client/client.go:359]: Agree with the comment above but this line needs to be refactored anyway after #372 is merged and we'll be no longer handling error here.

File: pkg/client/client.go (hunk 2 above)
Comment (pdyraga) [reply pkg/client/client.go:366]: ☝️

File: pkg/client/client.go (hunk 1 above)
Comment (pdyraga) [reply pkg/client/client.go:379]: ☝️

File: pkg/client/client.go [hunk 6]
```diff
@@ -425,18 +531,18 @@ func monitorKeepTerminatedEvent(
...
 // a provided function returning a boolean value.
-func waitForEventConfirmation(
+func waitForChainConfirmation(
 	ethereumChain eth.Handle,
 	startBlockNumber uint64,
 	stateCheck func() (bool, error),
 ) (bool, error) {
 	blockHeight := startBlockNumber + blockConfirmations
-	logger.Infof("waiting for [%d] block", blockHeight)
+	logger.Infof("waiting for block [%d]", blockHeight)
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/client/client.go:552]: Agreed. And I think `to confirm chain state` is even better.

**Pull request merged**